// playwright/lib/rpc/server/browserTypeDispatcher.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserTypeDispatcher = void 0;
const browserDispatcher_1 = require("./browserDispatcher");
const dispatcher_1 = require("./dispatcher");
const browserContextDispatcher_1 = require("./browserContextDispatcher");
const browserServerDispatcher_1 = require("./browserServerDispatcher");
const converters_1 = require("../../converters");
const jsHandleDispatcher_1 = require("./jsHandleDispatcher");
class BrowserTypeDispatcher extends dispatcher_1.Dispatcher {
    constructor(scope, browserType) {
        super(scope, browserType, 'BrowserType', {
            executablePath: browserType.executablePath(),
            name: browserType.name()
        }, true);
    }
    async launch(params) {
        const options = {
            ...params,
            ignoreDefaultArgs: params.ignoreAllDefaultArgs ? true : params.ignoreDefaultArgs,
            env: params.env ? converters_1.envArrayToObject(params.env) : undefined,
            firefoxUserPrefs: params.firefoxUserPrefs ? jsHandleDispatcher_1.parseValue(params.firefoxUserPrefs) : undefined,
        };
        const browser = await this._object.launch(options);
        return { browser: new browserDispatcher_1.BrowserDispatcher(this._scope, browser) };
    }
    async launchPersistentContext(params) {
        const options = {
            ...params,
            viewport: params.viewport || (params.noDefaultViewport ? null : undefined),
            ignoreDefaultArgs: params.ignoreAllDefaultArgs ? true : params.ignoreDefaultArgs,
            env: params.env ? converters_1.envArrayToObject(params.env) : undefined,
            extraHTTPHeaders: params.extraHTTPHeaders ? converters_1.headersArrayToObject(params.extraHTTPHeaders) : undefined,
        };
        const browserContext = await this._object.launchPersistentContext(params.userDataDir, options);
        return { context: new browserContextDispatcher_1.BrowserContextDispatcher(this._scope, browserContext) };
    }
    async launchServer(params) {
        const options = {
            ...params,
            ignoreDefaultArgs: params.ignoreAllDefaultArgs ? true : params.ignoreDefaultArgs,
            firefoxUserPrefs: params.firefoxUserPrefs ? jsHandleDispatcher_1.parseValue(params.firefoxUserPrefs) : undefined,
            env: params.env ? converters_1.envArrayToObject(params.env) : undefined,
        };
        return { server: new browserServerDispatcher_1.BrowserServerDispatcher(this._scope, await this._object.launchServer(options)) };
    }
    async connect(params) {
        const browser = await this._object.connect(params);
        return { browser: new browserDispatcher_1.BrowserDispatcher(this._scope, browser) };
    }
}
exports.BrowserTypeDispatcher = BrowserTypeDispatcher;
