// playwright/lib/rpc/client/browser.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Browser = void 0;
const browserContext_1 = require("./browserContext");
const channelOwner_1 = require("./channelOwner");
const events_1 = require("./events");
const converters_1 = require("../../converters");
class Browser extends channelOwner_1.ChannelOwner {
    constructor(parent, type, guid, initializer) {
        super(parent, type, guid, initializer);
        this._contexts = new Set();
        this._isConnected = true;
        this._isClosedOrClosing = false;
        this._browserType = parent;
        this._channel.on('close', () => {
            this._isConnected = false;
            this.emit(events_1.Events.Browser.Disconnected);
            this._isClosedOrClosing = true;
        });
        this._closedPromise = new Promise(f => this.once(events_1.Events.Browser.Disconnected, f));
    }
    static from(browser) {
        return browser._object;
    }
    static fromNullable(browser) {
        return browser ? Browser.from(browser) : null;
    }
    async newContext(options = {}) {
        const logger = options.logger;
        options = { ...options, logger: undefined };
        return this._wrapApiCall('browser.newContext', async () => {
            const contextOptions = {
                ...options,
                viewport: options.viewport === null ? undefined : options.viewport,
                noDefaultViewport: options.viewport === null,
                extraHTTPHeaders: options.extraHTTPHeaders ? converters_1.headersObjectToArray(options.extraHTTPHeaders) : undefined,
            };
            const context = browserContext_1.BrowserContext.from((await this._channel.newContext(contextOptions)).context);
            this._contexts.add(context);
            context._logger = logger || this._logger;
            return context;
        });
    }
    contexts() {
        return [...this._contexts];
    }
    version() {
        return this._initializer.version;
    }
    async newPage(options = {}) {
        const context = await this.newContext(options);
        const page = await context.newPage();
        page._ownedContext = context;
        context._ownerPage = page;
        return page;
    }
    isConnected() {
        return this._isConnected;
    }
    async close() {
        return this._wrapApiCall('browser.close', async () => {
            if (!this._isClosedOrClosing) {
                this._isClosedOrClosing = true;
                await this._channel.close();
            }
            await this._closedPromise;
        });
    }
}
exports.Browser = Browser;
