// playwright/lib/rpc/client/browserServer.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserServer = void 0;
const channelOwner_1 = require("./channelOwner");
const events_1 = require("./events");
class BrowserServer extends channelOwner_1.ChannelOwner {
    static from(server) {
        return server._object;
    }
    constructor(parent, type, guid, initializer) {
        super(parent, type, guid, initializer);
        this._channel.on('close', ({ exitCode, signal }) => {
            this.emit(events_1.Events.BrowserServer.Close, exitCode === undefined ? null : exitCode, signal === undefined ? null : signal);
        });
    }
    process() {
        return { pid: this._initializer.pid };
    }
    wsEndpoint() {
        return this._initializer.wsEndpoint;
    }
    async kill() {
        return this._wrapApiCall('browserServer.kill', async () => {
            await this._channel.kill();
        });
    }
    async close() {
        return this._wrapApiCall('browserServer.close', async () => {
            await this._channel.close();
        });
    }
    _checkLeaks() { }
}
exports.BrowserServer = BrowserServer;
