// playwright/lib/rpc/client/consoleMessage.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleMessage = void 0;
const util = require("util");
const jsHandle_1 = require("./jsHandle");
const channelOwner_1 = require("./channelOwner");
let __dummyInitializer;
class ConsoleMessage extends channelOwner_1.ChannelOwner {
    static from(message) {
        return message._object;
    }
    constructor(parent, type, guid, initializer) {
        super(parent, type, guid, initializer);
    }
    type() {
        return this._initializer.type;
    }
    text() {
        return this._initializer.text;
    }
    args() {
        return this._initializer.args.map(jsHandle_1.JSHandle.from);
    }
    location() {
        return this._initializer.location;
    }
    [util.inspect.custom]() {
        return this.text();
    }
}
exports.ConsoleMessage = ConsoleMessage;
