// playwright/lib/server/chromium.js
"use strict";
/**
 * Copyright 2017 Google Inc. All rights reserved.
 * Modifications copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Chromium = void 0;
const path = require("path");
const os = require("os");
const helper_1 = require("../helper");
const crBrowser_1 = require("../chromium/crBrowser");
const ws = require("ws");
const crConnection_1 = require("../chromium/crConnection");
const stackTrace_1 = require("../utils/stackTrace");
const browserType_1 = require("./browserType");
const crDevTools_1 = require("../chromium/crDevTools");
const webSocketServer_1 = require("./webSocketServer");
class Chromium extends browserType_1.BrowserTypeBase {
    constructor(packagePath, browser) {
        const debugPortStr = helper_1.getFromENV('PLAYWRIGHT_CHROMIUM_DEBUG_PORT');
        const debugPort = debugPortStr ? +debugPortStr : undefined;
        if (debugPort !== undefined) {
            if (Number.isNaN(debugPort))
                throw new Error(`PLAYWRIGHT_CHROMIUM_DEBUG_PORT must be a number, but is set to "${debugPortStr}"`);
            helper_1.logPolitely(`NOTE: Chromium will be launched in debug mode on port ${debugPort}`);
        }
        super(packagePath, browser, debugPort ? { webSocketRegex: /^DevTools listening on (ws:\/\/.*)$/, stream: 'stderr' } : null);
        this._debugPort = debugPort;
        if (helper_1.helper.isDebugMode())
            this._devtools = this._createDevTools();
    }
    _createDevTools() {
        return new crDevTools_1.CRDevTools(path.join(this._browserPath, 'devtools-preferences.json'));
    }
    async _connectToTransport(transport, options) {
        let devtools = this._devtools;
        if (options.__testHookForDevTools) {
            devtools = this._createDevTools();
            await options.__testHookForDevTools(devtools);
        }
        return crBrowser_1.CRBrowser.connect(transport, options, devtools);
    }
    _rewriteStartupError(error, prefix) {
        // These error messages are taken from Chromium source code as of July, 2020:
        // https://github.com/chromium/chromium/blob/70565f67e79f79e17663ad1337dc6e63ee207ce9/content/browser/zygote_host/zygote_host_impl_linux.cc
        if (!error.message.includes('crbug.com/357670') && !error.message.includes('No usable sandbox!') && !error.message.includes('crbug.com/638180'))
            return error;
        return stackTrace_1.rewriteErrorMessage(error, [
            `${prefix}: Chromium sandboxing failed!`,
            `================================`,
            `To workaround sandboxing issues, do either of the following:`,
            `  - (preferred): Configure environment to support sandboxing: https://github.com/microsoft/playwright/blob/master/docs/troubleshooting.md`,
            `  - (alternative): Launch Chromium without sandbox using 'chromiumSandbox: false' option`,
            `================================`,
            ``,
        ].join('\n'));
    }
    _amendEnvironment(env, userDataDir, executable, browserArguments) {
        return env;
    }
    _amendArguments(browserArguments) {
        // We currently only support Linux.
        if (os.platform() !== 'linux')
            return browserArguments;
        // If there's already --no-sandbox passed in, do nothing.
        if (browserArguments.indexOf('--no-sandbox') !== -1)
            return browserArguments;
        const runningAsRoot = process.geteuid && process.geteuid() === 0;
        if (runningAsRoot) {
            console.warn('WARNING: Playwright is being run under "root" user - disabling Chromium sandbox! Run under regular user to get rid of this warning.');
            return ['--no-sandbox', ...browserArguments];
        }
        return browserArguments;
    }
    _attemptToGracefullyCloseBrowser(transport) {
        const message = { method: 'Browser.close', id: crConnection_1.kBrowserCloseMessageId, params: {} };
        transport.send(message);
    }
    _startWebSocketServer(transport, logger, port) {
        return startWebSocketServer(transport, logger, port);
    }
    _defaultArgs(options, isPersistent, userDataDir) {
        const { args = [], proxy } = options;
        const userDataDirArg = args.find(arg => arg.startsWith('--user-data-dir'));
        if (userDataDirArg)
            throw new Error('Pass userDataDir parameter instead of specifying --user-data-dir argument');
        if (args.find(arg => arg.startsWith('--remote-debugging-pipe')))
            throw new Error('Playwright manages remote debugging connection itself.');
        if (args.find(arg => !arg.startsWith('-')))
            throw new Error('Arguments can not specify page to be opened');
        const chromeArguments = [...DEFAULT_ARGS];
        chromeArguments.push(`--user-data-dir=${userDataDir}`);
        if (this._debugPort !== undefined)
            chromeArguments.push('--remote-debugging-port=' + this._debugPort);
        else
            chromeArguments.push('--remote-debugging-pipe');
        if (options.devtools)
            chromeArguments.push('--auto-open-devtools-for-tabs');
        if (options.headless) {
            chromeArguments.push('--headless', '--hide-scrollbars', '--mute-audio');
        }
        if (options.chromiumSandbox === false)
            chromeArguments.push('--no-sandbox');
        if (proxy) {
            const proxyURL = new URL(proxy.server);
            const isSocks = proxyURL.protocol === 'socks5:';
            // https://www.chromium.org/developers/design-documents/network-settings
            if (isSocks) {
                // https://www.chromium.org/developers/design-documents/network-stack/socks-proxy
                chromeArguments.push(`--host-resolver-rules="MAP * ~NOTFOUND , EXCLUDE ${proxyURL.hostname}"`);
            }
            chromeArguments.push(`--proxy-server=${proxy.server}`);
            if (proxy.bypass) {
                const patterns = proxy.bypass.split(',').map(t => t.trim()).map(t => t.startsWith('.') ? '*' + t : t);
                chromeArguments.push(`--proxy-bypass-list=${patterns.join(';')}`);
            }
        }
        chromeArguments.push(...args);
        if (isPersistent)
            chromeArguments.push('about:blank');
        else
            chromeArguments.push('--no-startup-window');
        return chromeArguments;
    }
}
exports.Chromium = Chromium;
function startWebSocketServer(transport, logger, port) {
    const awaitingBrowserTarget = new Map();
    const sessionToData = new Map();
    const socketToBrowserSession = new Map();
    function addSession(sessionId, socket, parentSessionId) {
        sessionToData.set(sessionId, {
            socket,
            children: new Set(),
            isBrowserSession: !parentSessionId,
            parent: parentSessionId,
        });
        if (parentSessionId)
            sessionToData.get(parentSessionId).children.add(sessionId);
    }
    function removeSession(sessionId) {
        const data = sessionToData.get(sessionId);
        for (const child of data.children)
            removeSession(child);
        if (data.parent)
            sessionToData.get(data.parent).children.delete(sessionId);
        sessionToData.delete(sessionId);
    }
    const server = new webSocketServer_1.WebSocketServer(transport, logger, port, {
        onBrowserResponse(seqNum, source, message) {
            if (awaitingBrowserTarget.has(seqNum)) {
                const freshSocket = awaitingBrowserTarget.get(seqNum);
                awaitingBrowserTarget.delete(seqNum);
                const sessionId = message.result.sessionId;
                if (freshSocket.readyState !== ws.CLOSED && freshSocket.readyState !== ws.CLOSING) {
                    const { queue } = socketToBrowserSession.get(freshSocket);
                    for (const item of queue) {
                        item.sessionId = sessionId;
                        server.sendMessageToBrowser(item, source);
                    }
                    socketToBrowserSession.set(freshSocket, { sessionId });
                    addSession(sessionId, freshSocket);
                }
                else {
                    server.sendMessageToBrowserOneWay('Target.detachFromTarget', { sessionId });
                    socketToBrowserSession.delete(freshSocket);
                }
                return;
            }
            if (message.id === -1)
                return;
            // At this point everything we care about has sessionId.
            if (!message.sessionId)
                return;
            const data = sessionToData.get(message.sessionId);
            if (data && data.socket.readyState !== ws.CLOSING) {
                if (data.isBrowserSession)
                    delete message.sessionId;
                data.socket.send(JSON.stringify(message));
            }
        },
        onBrowserNotification(message) {
            // At this point everything we care about has sessionId.
            if (!message.sessionId)
                return;
            const data = sessionToData.get(message.sessionId);
            if (data && data.socket.readyState !== ws.CLOSING) {
                if (message.method === 'Target.attachedToTarget')
                    addSession(message.params.sessionId, data.socket, message.sessionId);
                if (message.method === 'Target.detachedFromTarget')
                    removeSession(message.params.sessionId);
                // Strip session ids from the browser sessions.
                if (data.isBrowserSession)
                    delete message.sessionId;
                data.socket.send(JSON.stringify(message));
            }
        },
        onClientAttached(socket) {
            socketToBrowserSession.set(socket, { queue: [] });
            const seqNum = server.sendMessageToBrowser({
                id: -1,
                method: 'Target.attachToBrowserTarget',
                params: {}
            }, socket);
            awaitingBrowserTarget.set(seqNum, socket);
        },
        onClientRequest(socket, message) {
            // If message has sessionId, pass through.
            if (message.sessionId) {
                server.sendMessageToBrowser(message, socket);
                return;
            }
            // If message has no sessionId, look it up.
            const session = socketToBrowserSession.get(socket);
            if (session.sessionId) {
                // We have it, use it.
                message.sessionId = session.sessionId;
                server.sendMessageToBrowser(message, socket);
                return;
            }
            // Pending session id, queue the message.
            session.queue.push(message);
        },
        onClientDetached(socket) {
            const session = socketToBrowserSession.get(socket);
            if (!session || !session.sessionId)
                return;
            removeSession(session.sessionId);
            socketToBrowserSession.delete(socket);
            server.sendMessageToBrowserOneWay('Target.detachFromTarget', { sessionId: session.sessionId });
        }
    });
    return server;
}
const DEFAULT_ARGS = [
    '--disable-background-networking',
    '--enable-features=NetworkService,NetworkServiceInProcess',
    '--disable-background-timer-throttling',
    '--disable-backgrounding-occluded-windows',
    '--disable-breakpad',
    '--disable-client-side-phishing-detection',
    '--disable-component-extensions-with-background-pages',
    '--disable-default-apps',
    '--disable-dev-shm-usage',
    '--disable-extensions',
    // BlinkGenPropertyTrees disabled due to crbug.com/937609
    '--disable-features=TranslateUI,BlinkGenPropertyTrees,ImprovedCookieControls,SameSiteByDefaultCookies',
    '--disable-hang-monitor',
    '--disable-ipc-flooding-protection',
    '--disable-popup-blocking',
    '--disable-prompt-on-repost',
    '--disable-renderer-backgrounding',
    '--disable-sync',
    '--force-color-profile=srgb',
    '--metrics-recording-only',
    '--no-first-run',
    '--enable-automation',
    '--password-store=basic',
    '--use-mock-keychain',
];
