// playwright/lib/server/webkit.js
"use strict";
/**
 * Copyright 2017 Google Inc. All rights reserved.
 * Modifications copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebKit = void 0;
const wkBrowser_1 = require("../webkit/wkBrowser");
const path = require("path");
const wkConnection_1 = require("../webkit/wkConnection");
const browserType_1 = require("./browserType");
const ws = require("ws");
const webSocketServer_1 = require("./webSocketServer");
const helper_1 = require("../helper");
class WebKit extends browserType_1.BrowserTypeBase {
    constructor(packagePath, browser) {
        super(packagePath, browser, null /* use pipe not websocket */);
    }
    _connectToTransport(transport, options) {
        return wkBrowser_1.WKBrowser.connect(transport, options);
    }
    _amendEnvironment(env, userDataDir, executable, browserArguments) {
        return { ...env, CURL_COOKIE_JAR_PATH: path.join(userDataDir, 'cookiejar.db') };
    }
    _amendArguments(browserArguments) {
        return browserArguments;
    }
    _rewriteStartupError(error, prefix) {
        return error;
    }
    _attemptToGracefullyCloseBrowser(transport) {
        transport.send({ method: 'Playwright.close', params: {}, id: wkConnection_1.kBrowserCloseMessageId });
    }
    _startWebSocketServer(transport, logger, port) {
        return startWebSocketServer(transport, logger, port);
    }
    _defaultArgs(options, isPersistent, userDataDir) {
        const { args = [], proxy, devtools, headless } = options;
        if (devtools)
            console.warn('devtools parameter as a launch argument in WebKit is not supported. Also starting Web Inspector manually will terminate the execution in WebKit.');
        const userDataDirArg = args.find(arg => arg.startsWith('--user-data-dir='));
        if (userDataDirArg)
            throw new Error('Pass userDataDir parameter instead of specifying --user-data-dir argument');
        if (args.find(arg => !arg.startsWith('-')))
            throw new Error('Arguments can not specify page to be opened');
        const webkitArguments = ['--inspector-pipe'];
        if (headless)
            webkitArguments.push('--headless');
        if (isPersistent)
            webkitArguments.push(`--user-data-dir=${userDataDir}`);
        else
            webkitArguments.push(`--no-startup-window`);
        if (proxy) {
            if (process.platform === 'darwin') {
                webkitArguments.push(`--proxy=${proxy.server}`);
                if (proxy.bypass)
                    webkitArguments.push(`--proxy-bypass-list=${proxy.bypass}`);
            }
            else if (process.platform === 'linux') {
                webkitArguments.push(`--proxy=${proxy.server}`);
                if (proxy.bypass)
                    webkitArguments.push(...proxy.bypass.split(',').map(t => `--ignore-host=${t}`));
            }
            else if (process.platform === 'win32') {
                webkitArguments.push(`--curl-proxy=${proxy.server}`);
                if (proxy.bypass)
                    webkitArguments.push(`--curl-noproxy=${proxy.bypass}`);
            }
        }
        webkitArguments.push(...args);
        if (isPersistent)
            webkitArguments.push('about:blank');
        return webkitArguments;
    }
}
exports.WebKit = WebKit;
function startWebSocketServer(transport, logger, port) {
    const pendingBrowserContextCreations = new Set();
    const pendingBrowserContextDeletions = new Map();
    const browserContextIds = new Map();
    const server = new webSocketServer_1.WebSocketServer(transport, logger, port, {
        onBrowserResponse(seqNum, source, message) {
            if (source.readyState === ws.CLOSED || source.readyState === ws.CLOSING) {
                if (pendingBrowserContextCreations.has(seqNum))
                    server.sendMessageToBrowserOneWay('Playwright.deleteContext', { browserContextId: message.result.browserContextId });
                return;
            }
            if (pendingBrowserContextCreations.has(seqNum)) {
                // Browser.createContext response -> establish context attribution.
                browserContextIds.set(message.result.browserContextId, source);
                pendingBrowserContextCreations.delete(seqNum);
            }
            const deletedContextId = pendingBrowserContextDeletions.get(seqNum);
            if (deletedContextId) {
                // Browser.deleteContext response -> remove context attribution.
                browserContextIds.delete(deletedContextId);
                pendingBrowserContextDeletions.delete(seqNum);
            }
            source.send(JSON.stringify(message));
            return;
        },
        onBrowserNotification(message) {
            // Process notification response.
            const { params, browserContextId } = message;
            const contextId = browserContextId || params.browserContextId;
            helper_1.assert(contextId);
            const socket = browserContextIds.get(contextId);
            if (!socket || socket.readyState === ws.CLOSING) {
                // Drop unattributed messages on the floor.
                return;
            }
            socket.send(JSON.stringify(message));
        },
        onClientAttached(socket) {
        },
        onClientRequest(socket, message) {
            const { method, params } = message;
            const seqNum = server.sendMessageToBrowser(message, socket);
            if (method === 'Playwright.createContext')
                pendingBrowserContextCreations.add(seqNum);
            if (method === 'Playwright.deleteContext')
                pendingBrowserContextDeletions.set(seqNum, params.browserContextId);
        },
        onClientDetached(socket) {
            for (const [browserContextId, s] of browserContextIds) {
                if (s === socket) {
                    server.sendMessageToBrowserOneWay('Playwright.deleteContext', { browserContextId });
                    browserContextIds.delete(browserContextId);
                }
            }
        }
    });
    return server;
}
