// playwright/lib/api.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var accessibility_1 = require("./accessibility");
Object.defineProperty(exports, "Accessibility", { enumerable: true, get: function () { return accessibility_1.Accessibility; } });
var console_1 = require("./console");
Object.defineProperty(exports, "ConsoleMessage", { enumerable: true, get: function () { return console_1.ConsoleMessage; } });
var dialog_1 = require("./dialog");
Object.defineProperty(exports, "Dialog", { enumerable: true, get: function () { return dialog_1.Dialog; } });
var download_1 = require("./download");
Object.defineProperty(exports, "Download", { enumerable: true, get: function () { return download_1.Download; } });
var dom_1 = require("./dom");
Object.defineProperty(exports, "ElementHandle", { enumerable: true, get: function () { return dom_1.ElementHandle; } });
var fileChooser_1 = require("./fileChooser");
Object.defineProperty(exports, "FileChooser", { enumerable: true, get: function () { return fileChooser_1.FileChooser; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "TimeoutError", { enumerable: true, get: function () { return errors_1.TimeoutError; } });
var frames_1 = require("./frames");
Object.defineProperty(exports, "Frame", { enumerable: true, get: function () { return frames_1.Frame; } });
var input_1 = require("./input");
Object.defineProperty(exports, "Keyboard", { enumerable: true, get: function () { return input_1.Keyboard; } });
Object.defineProperty(exports, "Mouse", { enumerable: true, get: function () { return input_1.Mouse; } });
var javascript_1 = require("./javascript");
Object.defineProperty(exports, "JSHandle", { enumerable: true, get: function () { return javascript_1.JSHandle; } });
var network_1 = require("./network");
Object.defineProperty(exports, "Request", { enumerable: true, get: function () { return network_1.Request; } });
Object.defineProperty(exports, "Response", { enumerable: true, get: function () { return network_1.Response; } });
Object.defineProperty(exports, "Route", { enumerable: true, get: function () { return network_1.Route; } });
var page_1 = require("./page");
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return page_1.Page; } });
Object.defineProperty(exports, "Worker", { enumerable: true, get: function () { return page_1.Worker; } });
var selectors_1 = require("./selectors");
Object.defineProperty(exports, "Selectors", { enumerable: true, get: function () { return selectors_1.Selectors; } });
var crBrowser_1 = require("./chromium/crBrowser");
Object.defineProperty(exports, "ChromiumBrowser", { enumerable: true, get: function () { return crBrowser_1.CRBrowser; } });
var crBrowser_2 = require("./chromium/crBrowser");
Object.defineProperty(exports, "ChromiumBrowserContext", { enumerable: true, get: function () { return crBrowser_2.CRBrowserContext; } });
var crCoverage_1 = require("./chromium/crCoverage");
Object.defineProperty(exports, "ChromiumCoverage", { enumerable: true, get: function () { return crCoverage_1.CRCoverage; } });
var crConnection_1 = require("./chromium/crConnection");
Object.defineProperty(exports, "CDPSession", { enumerable: true, get: function () { return crConnection_1.CRSession; } });
var ffBrowser_1 = require("./firefox/ffBrowser");
Object.defineProperty(exports, "FirefoxBrowser", { enumerable: true, get: function () { return ffBrowser_1.FFBrowser; } });
var wkBrowser_1 = require("./webkit/wkBrowser");
Object.defineProperty(exports, "WebKitBrowser", { enumerable: true, get: function () { return wkBrowser_1.WKBrowser; } });
var browserServer_1 = require("./server/browserServer");
Object.defineProperty(exports, "BrowserServer", { enumerable: true, get: function () { return browserServer_1.BrowserServer; } });
