// playwright/lib/dialog.js
"use strict";
/**
 * Copyright 2017 Google Inc. All rights reserved.
 * Modifications copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dialog = void 0;
const helper_1 = require("./helper");
class Dialog {
    constructor(logger, type, message, onHandle, defaultValue) {
        this._handled = false;
        this._logger = logger;
        this._type = type;
        this._message = message;
        this._onHandle = onHandle;
        this._defaultValue = defaultValue || '';
        this._logger.info(`  ${this._preview()} was shown`);
    }
    type() {
        return this._type;
    }
    message() {
        return this._message;
    }
    defaultValue() {
        return this._defaultValue;
    }
    async accept(promptText) {
        helper_1.assert(!this._handled, 'Cannot accept dialog which is already handled!');
        this._handled = true;
        this._logger.info(`  ${this._preview()} was accepted`);
        await this._onHandle(true, promptText);
    }
    async dismiss() {
        helper_1.assert(!this._handled, 'Cannot dismiss dialog which is already handled!');
        this._handled = true;
        this._logger.info(`  ${this._preview()} was dismissed`);
        await this._onHandle(false);
    }
    _preview() {
        if (!this._message)
            return this._type;
        if (this._message.length <= 50)
            return `${this._type} "${this._message}"`;
        return `${this._type} "${this._message.substring(0, 49) + '\u2026'}"`;
    }
}
exports.Dialog = Dialog;
