// playwright/lib/page.js
"use strict";
/**
 * Copyright 2017 Google Inc. All rights reserved.
 * Modifications copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageBinding = exports.Worker = exports.Page = void 0;
const frames = require("./frames");
const helper_1 = require("./helper");
const input = require("./input");
const js = require("./javascript");
const network = require("./network");
const screenshotter_1 = require("./screenshotter");
const timeoutSettings_1 = require("./timeoutSettings");
const types = require("./types");
const events_1 = require("./events");
const console_1 = require("./console");
const accessibility = require("./accessibility");
const events_2 = require("events");
const fileChooser_1 = require("./fileChooser");
const logger_1 = require("./logger");
const progress_1 = require("./progress");
class Page extends events_2.EventEmitter {
    constructor(delegate, browserContext) {
        super();
        this._closedState = 'open';
        this._disconnected = false;
        this._pageBindings = new Map();
        this._evaluateOnNewDocumentSources = [];
        this._workers = new Map();
        this._routes = [];
        this._callingPageAPI = false;
        this._delegate = delegate;
        this._logger = browserContext._apiLogger;
        this._closedCallback = () => { };
        this._closedPromise = new Promise(f => this._closedCallback = f);
        this._disconnectedCallback = () => { };
        this._disconnectedPromise = new Promise(f => this._disconnectedCallback = f);
        this._crashedCallback = () => { };
        this._crashedPromise = new Promise(f => this._crashedCallback = f);
        this._browserContext = browserContext;
        this._state = {
            viewportSize: browserContext._options.viewport ? { ...browserContext._options.viewport } : null,
            mediaType: null,
            colorScheme: null,
            extraHTTPHeaders: null,
        };
        this.accessibility = new accessibility.Accessibility(delegate.getAccessibilityTree.bind(delegate));
        this.keyboard = new input.Keyboard(delegate.rawKeyboard);
        this.mouse = new input.Mouse(delegate.rawMouse, this.keyboard);
        this._timeoutSettings = new timeoutSettings_1.TimeoutSettings(browserContext._timeoutSettings);
        this._screenshotter = new screenshotter_1.Screenshotter(this);
        this._frameManager = new frames.FrameManager(this);
        if (delegate.pdf)
            this.pdf = delegate.pdf.bind(delegate);
        this.coverage = delegate.coverage ? delegate.coverage() : null;
    }
    _didClose() {
        this._frameManager.dispose();
        helper_1.assert(this._closedState !== 'closed', 'Page closed twice');
        this._closedState = 'closed';
        this.emit(events_1.Events.Page.Close);
        this._closedCallback();
    }
    _didCrash() {
        this._frameManager.dispose();
        this.emit(events_1.Events.Page.Crash);
        this._crashedCallback(new Error('Page crashed'));
    }
    _didDisconnect() {
        this._frameManager.dispose();
        helper_1.assert(!this._disconnected, 'Page disconnected twice');
        this._disconnected = true;
        this._disconnectedCallback(new Error('Page closed'));
    }
    async _runAbortableTask(task, timeout, apiName) {
        return progress_1.runAbortableTask(async (progress) => {
            return task(progress);
        }, this._logger, timeout, apiName);
    }
    async _onFileChooserOpened(handle) {
        const multiple = await handle.evaluate(element => !!element.multiple);
        if (!this.listenerCount(events_1.Events.Page.FileChooser)) {
            handle.dispose();
            return;
        }
        const fileChooser = new fileChooser_1.FileChooser(this, handle, multiple);
        this.emit(events_1.Events.Page.FileChooser, fileChooser);
    }
    context() {
        return this._browserContext;
    }
    async opener() {
        return await this._delegate.opener();
    }
    mainFrame() {
        return this._frameManager.mainFrame();
    }
    frame(options) {
        const name = helper_1.helper.isString(options) ? options : options.name;
        const url = helper_1.helper.isObject(options) ? options.url : undefined;
        helper_1.assert(name || url, 'Either name or url matcher should be specified');
        return this.frames().find(f => {
            if (name)
                return f.name() === name;
            return helper_1.helper.urlMatches(f.url(), url);
        }) || null;
    }
    frames() {
        return this._frameManager.frames();
    }
    setDefaultNavigationTimeout(timeout) {
        this._timeoutSettings.setDefaultNavigationTimeout(timeout);
    }
    setDefaultTimeout(timeout) {
        this._timeoutSettings.setDefaultTimeout(timeout);
    }
    async $(selector) {
        return this._attributeToPage(() => this.mainFrame().$(selector));
    }
    async waitForSelector(selector, options) {
        return this._attributeToPage(() => this.mainFrame().waitForSelector(selector, options));
    }
    async dispatchEvent(selector, type, eventInit, options) {
        return this._attributeToPage(() => this.mainFrame().dispatchEvent(selector, type, eventInit, options));
    }
    async evaluateHandle(pageFunction, arg) {
        helper_1.assertMaxArguments(arguments.length, 2);
        return this._attributeToPage(() => this.mainFrame().evaluateHandle(pageFunction, arg));
    }
    async _evaluateExpressionHandle(expression, isFunction, arg) {
        return this._attributeToPage(() => this.mainFrame()._evaluateExpressionHandle(expression, isFunction, arg));
    }
    async $eval(selector, pageFunction, arg) {
        helper_1.assertMaxArguments(arguments.length, 3);
        return this._attributeToPage(() => this.mainFrame().$eval(selector, pageFunction, arg));
    }
    async _$evalExpression(selector, expression, isFunction, arg) {
        return this._attributeToPage(() => this.mainFrame()._$evalExpression(selector, expression, isFunction, arg));
    }
    async $$eval(selector, pageFunction, arg) {
        helper_1.assertMaxArguments(arguments.length, 3);
        return this._attributeToPage(() => this.mainFrame().$$eval(selector, pageFunction, arg));
    }
    async _$$evalExpression(selector, expression, isFunction, arg) {
        return this._attributeToPage(() => this.mainFrame()._$$evalExpression(selector, expression, isFunction, arg));
    }
    async $$(selector) {
        return this._attributeToPage(() => this.mainFrame().$$(selector));
    }
    async addScriptTag(options) {
        return this._attributeToPage(() => this.mainFrame().addScriptTag(options));
    }
    async addStyleTag(options) {
        return this._attributeToPage(() => this.mainFrame().addStyleTag(options));
    }
    async exposeFunction(name, playwrightFunction) {
        await this.exposeBinding(name, (options, ...args) => playwrightFunction(...args));
    }
    async exposeBinding(name, playwrightBinding) {
        if (this._pageBindings.has(name))
            throw new Error(`Function "${name}" has been already registered`);
        if (this._browserContext._pageBindings.has(name))
            throw new Error(`Function "${name}" has been already registered in the browser context`);
        const binding = new PageBinding(name, playwrightBinding);
        this._pageBindings.set(name, binding);
        await this._delegate.exposeBinding(binding);
    }
    setExtraHTTPHeaders(headers) {
        this._state.extraHTTPHeaders = network.verifyHeaders(headers);
        return this._delegate.updateExtraHTTPHeaders();
    }
    async _onBindingCalled(payload, context) {
        if (this._disconnected || this._closedState === 'closed')
            return;
        await PageBinding.dispatch(this, payload, context);
    }
    _addConsoleMessage(type, args, location, text) {
        const message = new console_1.ConsoleMessage(type, text, args, location);
        const intercepted = this._frameManager.interceptConsoleMessage(message);
        if (intercepted || !this.listenerCount(events_1.Events.Page.Console))
            args.forEach(arg => arg.dispose());
        else
            this.emit(events_1.Events.Page.Console, message);
    }
    url() {
        return this._attributeToPage(() => this.mainFrame().url());
    }
    async content() {
        return this._attributeToPage(() => this.mainFrame().content());
    }
    async setContent(html, options) {
        return this._attributeToPage(() => this.mainFrame().setContent(html, options));
    }
    async goto(url, options) {
        return this._attributeToPage(() => this.mainFrame().goto(url, options));
    }
    async reload(options) {
        const waitPromise = this.waitForNavigation(options);
        await this._delegate.reload();
        return waitPromise;
    }
    async waitForLoadState(state, options) {
        return this._attributeToPage(() => this.mainFrame().waitForLoadState(state, options));
    }
    async waitForNavigation(options) {
        return this._attributeToPage(() => this.mainFrame().waitForNavigation(options));
    }
    async waitForRequest(urlOrPredicate, options = {}) {
        const predicate = (request) => {
            if (helper_1.helper.isString(urlOrPredicate) || helper_1.helper.isRegExp(urlOrPredicate))
                return helper_1.helper.urlMatches(request.url(), urlOrPredicate);
            return urlOrPredicate(request);
        };
        return this.waitForEvent(events_1.Events.Page.Request, { predicate, timeout: options.timeout });
    }
    async waitForResponse(urlOrPredicate, options = {}) {
        const predicate = (response) => {
            if (helper_1.helper.isString(urlOrPredicate) || helper_1.helper.isRegExp(urlOrPredicate))
                return helper_1.helper.urlMatches(response.url(), urlOrPredicate);
            return urlOrPredicate(response);
        };
        return this.waitForEvent(events_1.Events.Page.Response, { predicate, timeout: options.timeout });
    }
    async waitForEvent(event, optionsOrPredicate = {}) {
        const options = typeof optionsOrPredicate === 'function' ? { predicate: optionsOrPredicate } : optionsOrPredicate;
        const progressController = new progress_1.ProgressController(this._logger, this._timeoutSettings.timeout(options), 'page.waitForEvent');
        this._disconnectedPromise.then(error => progressController.abort(error));
        if (event !== events_1.Events.Page.Crash)
            this._crashedPromise.then(error => progressController.abort(error));
        return progressController.run(progress => helper_1.helper.waitForEvent(progress, this, event, options.predicate).promise);
    }
    async goBack(options) {
        const waitPromise = this.waitForNavigation(options);
        const result = await this._delegate.goBack();
        if (!result) {
            waitPromise.catch(() => { });
            return null;
        }
        return waitPromise;
    }
    async goForward(options) {
        const waitPromise = this.waitForNavigation(options);
        const result = await this._delegate.goForward();
        if (!result) {
            waitPromise.catch(() => { });
            return null;
        }
        return waitPromise;
    }
    async emulateMedia(options) {
        if (options.media !== undefined)
            helper_1.assert(options.media === null || types.mediaTypes.has(options.media), 'media: expected one of (screen|print|null)');
        if (options.colorScheme !== undefined)
            helper_1.assert(options.colorScheme === null || types.colorSchemes.has(options.colorScheme), 'colorScheme: expected one of (dark|light|no-preference|null)');
        if (options.media !== undefined)
            this._state.mediaType = options.media;
        if (options.colorScheme !== undefined)
            this._state.colorScheme = options.colorScheme;
        await this._delegate.updateEmulateMedia();
    }
    async setViewportSize(viewportSize) {
        this._state.viewportSize = { ...viewportSize };
        await this._delegate.setViewportSize(this._state.viewportSize);
    }
    viewportSize() {
        return this._state.viewportSize;
    }
    async bringToFront() {
        await this._delegate.bringToFront();
    }
    async evaluate(pageFunction, arg) {
        helper_1.assertMaxArguments(arguments.length, 2);
        return this._attributeToPage(() => this.mainFrame().evaluate(pageFunction, arg));
    }
    async _evaluateExpression(expression, isFunction, arg) {
        return this._attributeToPage(() => this.mainFrame()._evaluateExpression(expression, isFunction, arg));
    }
    async addInitScript(script, arg) {
        const source = await helper_1.helper.evaluationScript(script, arg);
        await this._addInitScriptExpression(source);
    }
    async _addInitScriptExpression(source) {
        this._evaluateOnNewDocumentSources.push(source);
        await this._delegate.evaluateOnNewDocument(source);
    }
    _needsRequestInterception() {
        return this._routes.length > 0 || this._browserContext._routes.length > 0;
    }
    async route(url, handler) {
        this._routes.push({ url, handler });
        await this._delegate.updateRequestInterception();
    }
    async unroute(url, handler) {
        this._routes = this._routes.filter(route => route.url !== url || (handler && route.handler !== handler));
        await this._delegate.updateRequestInterception();
    }
    _requestStarted(request) {
        this.emit(events_1.Events.Page.Request, request);
        const route = request._route();
        if (!route)
            return;
        for (const { url, handler } of this._routes) {
            if (helper_1.helper.urlMatches(request.url(), url)) {
                handler(route, request);
                return;
            }
        }
        for (const { url, handler } of this._browserContext._routes) {
            if (helper_1.helper.urlMatches(request.url(), url)) {
                handler(route, request);
                return;
            }
        }
        route.continue();
    }
    _isRouted(requestURL) {
        for (const { url } of this._routes) {
            if (helper_1.helper.urlMatches(requestURL, url))
                return true;
        }
        for (const { url } of this._browserContext._routes) {
            if (helper_1.helper.urlMatches(requestURL, url))
                return true;
        }
        return false;
    }
    async screenshot(options = {}) {
        return this._runAbortableTask(progress => this._screenshotter.screenshotPage(progress, options), this._timeoutSettings.timeout(options), 'page.screenshot');
    }
    async title() {
        return this._attributeToPage(() => this.mainFrame().title());
    }
    async close(options) {
        if (this._closedState === 'closed')
            return;
        const runBeforeUnload = !!options && !!options.runBeforeUnload;
        if (this._closedState !== 'closing') {
            this._closedState = 'closing';
            helper_1.assert(!this._disconnected, 'Protocol error: Connection closed. Most likely the page has been closed.');
            await this._delegate.closePage(runBeforeUnload);
        }
        if (!runBeforeUnload)
            await this._closedPromise;
        if (this._ownedContext)
            await this._ownedContext.close();
    }
    _setIsError() {
        if (!this._frameManager.mainFrame())
            this._frameManager.frameAttached('<dummy>', null);
    }
    isClosed() {
        return this._closedState === 'closed';
    }
    _attributeToPage(func) {
        try {
            this._callingPageAPI = true;
            return func();
        }
        finally {
            this._callingPageAPI = false;
        }
    }
    async click(selector, options) {
        return this._attributeToPage(() => this.mainFrame().click(selector, options));
    }
    async dblclick(selector, options) {
        return this._attributeToPage(() => this.mainFrame().dblclick(selector, options));
    }
    async fill(selector, value, options) {
        return this._attributeToPage(() => this.mainFrame().fill(selector, value, options));
    }
    async focus(selector, options) {
        return this._attributeToPage(() => this.mainFrame().focus(selector, options));
    }
    async textContent(selector, options) {
        return this._attributeToPage(() => this.mainFrame().textContent(selector, options));
    }
    async innerText(selector, options) {
        return this._attributeToPage(() => this.mainFrame().innerText(selector, options));
    }
    async innerHTML(selector, options) {
        return this._attributeToPage(() => this.mainFrame().innerHTML(selector, options));
    }
    async getAttribute(selector, name, options) {
        return this._attributeToPage(() => this.mainFrame().getAttribute(selector, name, options));
    }
    async hover(selector, options) {
        return this._attributeToPage(() => this.mainFrame().hover(selector, options));
    }
    async selectOption(selector, values, options) {
        return this._attributeToPage(() => this.mainFrame().selectOption(selector, values, options));
    }
    async setInputFiles(selector, files, options) {
        return this._attributeToPage(() => this.mainFrame().setInputFiles(selector, files, options));
    }
    async type(selector, text, options) {
        return this._attributeToPage(() => this.mainFrame().type(selector, text, options));
    }
    async press(selector, key, options) {
        return this._attributeToPage(() => this.mainFrame().press(selector, key, options));
    }
    async check(selector, options) {
        return this._attributeToPage(() => this.mainFrame().check(selector, options));
    }
    async uncheck(selector, options) {
        return this._attributeToPage(() => this.mainFrame().uncheck(selector, options));
    }
    async waitForTimeout(timeout) {
        await this.mainFrame().waitForTimeout(timeout);
    }
    async waitForFunction(pageFunction, arg, options) {
        return this._attributeToPage(() => this.mainFrame().waitForFunction(pageFunction, arg, options));
    }
    async _waitForFunctionExpression(expression, isFunction, arg, options = {}) {
        return this._attributeToPage(() => this.mainFrame()._waitForFunctionExpression(expression, isFunction, arg, options));
    }
    workers() {
        return [...this._workers.values()];
    }
    _addWorker(workerId, worker) {
        this._workers.set(workerId, worker);
        this.emit(events_1.Events.Page.Worker, worker);
    }
    _removeWorker(workerId) {
        const worker = this._workers.get(workerId);
        if (!worker)
            return;
        worker.emit(events_1.Events.Worker.Close, worker);
        this._workers.delete(workerId);
    }
    _clearWorkers() {
        for (const [workerId, worker] of this._workers) {
            worker.emit(events_1.Events.Worker.Close, worker);
            this._workers.delete(workerId);
        }
    }
    on(event, listener) {
        if (event === events_1.Events.Page.FileChooser) {
            if (!this.listenerCount(event))
                this._delegate.setFileChooserIntercepted(true);
        }
        super.on(event, listener);
        return this;
    }
    removeListener(event, listener) {
        super.removeListener(event, listener);
        if (event === events_1.Events.Page.FileChooser && !this.listenerCount(event))
            this._delegate.setFileChooserIntercepted(false);
        return this;
    }
}
exports.Page = Page;
class Worker extends events_2.EventEmitter {
    constructor(url) {
        super();
        this._existingExecutionContext = null;
        this._url = url;
        this._executionContextCallback = () => { };
        this._executionContextPromise = new Promise(x => this._executionContextCallback = x);
    }
    _createExecutionContext(delegate) {
        this._existingExecutionContext = new js.ExecutionContext(delegate);
        this._executionContextCallback(this._existingExecutionContext);
    }
    url() {
        return this._url;
    }
    async evaluate(pageFunction, arg) {
        helper_1.assertMaxArguments(arguments.length, 2);
        return js.evaluate(await this._executionContextPromise, true /* returnByValue */, pageFunction, arg);
    }
    async _evaluateExpression(expression, isFunction, arg) {
        return js.evaluateExpression(await this._executionContextPromise, true /* returnByValue */, expression, isFunction, arg);
    }
    async evaluateHandle(pageFunction, arg) {
        helper_1.assertMaxArguments(arguments.length, 2);
        return js.evaluate(await this._executionContextPromise, false /* returnByValue */, pageFunction, arg);
    }
    async _evaluateExpressionHandle(expression, isFunction, arg) {
        return js.evaluateExpression(await this._executionContextPromise, false /* returnByValue */, expression, isFunction, arg);
    }
}
exports.Worker = Worker;
class PageBinding {
    constructor(name, playwrightFunction) {
        this.name = name;
        this.playwrightFunction = playwrightFunction;
        this.source = helper_1.helper.evaluationString(addPageBinding, name);
    }
    static async dispatch(page, payload, context) {
        const { name, seq, args } = JSON.parse(payload);
        try {
            let binding = page._pageBindings.get(name);
            if (!binding)
                binding = page._browserContext._pageBindings.get(name);
            const result = await binding.playwrightFunction({ frame: context.frame, page, context: page._browserContext }, ...args);
            context.evaluateInternal(deliverResult, { name, seq, result }).catch(logger_1.logError(page._logger));
        }
        catch (error) {
            if (helper_1.helper.isError(error))
                context.evaluateInternal(deliverError, { name, seq, message: error.message, stack: error.stack }).catch(logger_1.logError(page._logger));
            else
                context.evaluateInternal(deliverErrorValue, { name, seq, error }).catch(logger_1.logError(page._logger));
        }
        function deliverResult(arg) {
            window[arg.name]['callbacks'].get(arg.seq).resolve(arg.result);
            window[arg.name]['callbacks'].delete(arg.seq);
        }
        function deliverError(arg) {
            const error = new Error(arg.message);
            error.stack = arg.stack;
            window[arg.name]['callbacks'].get(arg.seq).reject(error);
            window[arg.name]['callbacks'].delete(arg.seq);
        }
        function deliverErrorValue(arg) {
            window[arg.name]['callbacks'].get(arg.seq).reject(arg.error);
            window[arg.name]['callbacks'].delete(arg.seq);
        }
    }
}
exports.PageBinding = PageBinding;
function addPageBinding(bindingName) {
    const binding = window[bindingName];
    if (binding.__installed)
        return;
    window[bindingName] = (...args) => {
        const me = window[bindingName];
        let callbacks = me['callbacks'];
        if (!callbacks) {
            callbacks = new Map();
            me['callbacks'] = callbacks;
        }
        const seq = (me['lastSeq'] || 0) + 1;
        me['lastSeq'] = seq;
        const promise = new Promise((resolve, reject) => callbacks.set(seq, { resolve, reject }));
        binding(JSON.stringify({ name: bindingName, seq, args }));
        return promise;
    };
    window[bindingName].__installed = true;
}
